postinstall.js

```diff
@@ -1,15 +1,97 @@
+const NGC_BINARY = resolve('./node_modules/.bin/ngc');
+
+async function main() {
+  // Applying all the patches to the packages
+  await execCommand(`node ${require.resolve('patch-package')}`);
+  // when running with bazel we need to perform a ngcc to compile
+  // all dependencies for ivy. If we are not run by bazel just build
+  // the workspace library. With bazel this is already in the build cache.
+
+  // Generate Angular ngfactory.js, ngsummary.js files for the dependencies,
+  // that are needed for ViewEngine
+  await execCommand(`${NGC_BINARY} -p view-engine-tsconfig.json`);
+  // Generate Ivy entry points
+  await execCommand(
+    `${NGCC_BINARY} --properties es2015 browser module main --first-only --create-ivy-entry-points`,
+  );
+  // link the ivy entry points
+  updateNgccMainFields();
+
+  if (!process.env.BAZEL_NPM_INSTALL) {
+    await execCommand(`npm run ng build workspace`);
+  }
+}
+
+/**
+ * Executes a shell command and return it as a Promise.
+ * @param cmd {string}
+ * @return {Promise<string>}
+ */
+function execCommand(cmd) {
+  return new Promise((resolve, reject) => {
+    exec(cmd, (error, stdout, stderr) => {
+      if (error) {
+        return reject(error);
+      }
+      const output = stdout ? stdout : stderr;
+      console.log(output);
+      resolve(output);
+    });
+  });
+}
+
+/**
+ * Script that runs after node modules have been installed, and Ngcc processed all packages.
+ * This script updates the `package.json` files of Angular framework packages to point to the
+ * Ngcc processed UMD bundles. This is needed because we run Angular in a `nodejs_binary`, but
+ * want to make sure that Ivy is being used. By default, the NodeJS module resolution will load
+ * the unprocessed UMD bundle because the `main` field of the `package.json` files point to the
+ * View Engine UMD bundles. This script updates the `main` field in `package.json` files to point
+ * to the previously generated Ivy UMD bundles.
+ *
+ * Ngcc does not by edit the `main` field because we ran it with the `--create-ivy-entry-points`
+ * flag. It instructs Ngcc to not modify existing package bundles, but rather create separate
+ * copies with the needed Ivy modifications. This is necessary because the original bundles
+ * are needed for View Engine, and we want to preserve them in order to be able to switch
+ * between Ivy and View Engine (for testing). Since the goal of this flag is to not modify
+ * any original package files/bundles, Ngcc will not edit the `main` field to point to
+ * the processed Ivy bundles.
+ */
+function updateNgccMainFields() {
+  sync('node_modules/@angular/**/package.json').forEach((filePath) => {
+    // Do not update `package.json` files for deeply nested node modules (e.g. dependencies of
+    // the `@angular/compiler-cli` package).
+    if (filePath.lastIndexOf('node_modules/') !== 0) {
+      return;
+    }
+    const parsedJson = JSON.parse(readFileSync(filePath, 'utf8'));
```

Comment:
A lot could go wrong here, invalid json the fields not being there etc. Does it make sense wrap this in a try catch and handle a bit more gracefully?

Reply:
This is only a parse for the angulars package.json if this one is broken all angular apps would be broken so I cannot see any benefit of adding a try catch here.

Reply:
For reference where I copied the ngcc madness that have to be done: https://github.com/angular/components/blob/master/tools/postinstall/update-ngcc-main-fields.js#L31